rust/src/client/experimental/client.rs

```diff
@@ -0,0 +1,299 @@
+            round_params,
+        } = self;
+
+        match participant.determine_role(
+            round_params.seed.as_slice(),
+            round_params.sum,
+            round_params.update,
+        ) {
+            Role::Unselected(participant) => {
+                info!("unselected");
+                ClientState::<Awaiting>::new(participant.reset(), round_params).into()
+            }
+            Role::Summer(participant) => ClientState::<Sum>::new(participant, round_params).into(),
+            Role::Updater(participant) => {
+                ClientState::<Update>::new(participant, round_params).into()
+            }
+        }
+    }
+}
+
+impl ClientState<Sum> {
+    fn new(participant: Participant<Sum>, round_params: RoundParameters) -> Self {
+        Self {
+            participant,
+            round_params,
+        }
+    }
+
+    async fn next(mut self, proxy: &mut Proxy) -> ClientStateMachine {
+        info!("selected to sum");
+
+        match self.run(proxy).await {
+            Ok(_) => self.into_sum2().into(),
+            Err(ClientError::RoundOutdated) => self.reset().into(),
+            Err(err) => {
+                error!("{:?}", err);
+                self.into()
+            }
+        }
+    }
+
+    async fn run(&mut self, proxy: &mut Proxy) -> Result<(), ClientError> {
+        self.check_round_freshness(proxy).await?;
+
+        let sum_msg = self.participant.compose_sum_message(&self.round_params.pk);
+        let sealed_msg = self
+            .participant
+            .seal_message(&self.round_params.pk, &sum_msg);
+
+        debug!("sending sum message");
+        proxy.post_message(sealed_msg).await?;
+        debug!("sum message sent");
+        Ok(())
+    }
+
+    fn into_sum2(self) -> ClientState<Sum2> {
+        ClientState::<Sum2>::new(self.participant.into(), self.round_params)
+    }
+}
+
+impl ClientState<Update> {
+    fn new(participant: Participant<Update>, round_params: RoundParameters) -> Self {
+        Self {
+            participant,
+            round_params,
+        }
+    }
+
+    async fn next<L: LocalModel>(
+        mut self,
+        proxy: &mut Proxy,
+        local_model: &mut L,
+    ) -> ClientStateMachine {
+        info!("selected to update");
+
+        match self.run(proxy, local_model).await {
+            Ok(_) | Err(ClientError::RoundOutdated) => self.reset().into(),
+            Err(err) => {
+                error!("{:?}", err);
+                self.into()
+            }
+        }
+    }
+
+    async fn run<L: LocalModel>(
+        &mut self,
+        proxy: &mut Proxy,
+        local_model: &mut L,
+    ) -> Result<(), ClientError> {
+        self.check_round_freshness(proxy).await?;
+
+        debug!("polling for local model");
+        let local_model = local_model
+            .get_local_model()
+            .await
+            .ok_or(ClientError::TooEarly("local model"))?;
+
+        debug!("polling for model scalar");
+        let scalar = proxy
```

Comment:
**Future improvments**

- cache received values. For example, if the request of the `sum_dict` fails and the task is tried again later, the `scalar` does not have to be queried again 
- the cache must be cleared after each new round